keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1055 @@
+> callee goroutine returns.
+
+
+With the `timeout` filter removed, we need to build a new mechanism to maintain parity.
+
+##### Global Synchronized Hashmap
+
+We can setup a `PostStartHook` that will detect how long a requst handler
+goroutine is active after the timeout happens, and provide the same observability
+as today.
+
+One simple approach for alpha, is to maintain a global synchronized hashmap of 
+the requests that are being tracked by the `WithPerRequestTimeout` filter.
+```
+type Tracker interface {
+	Enter(key *http.Request, data *RequestData)
+	Leave (key *http.Request)
+}
+
+type RequestData struct {
+	StartedAt      time.Time
+	DeadlineAt     time.Time
+	RequestURI     string
+	RequestAuditID string
+}
+
+type tracker struct{
+	inflight map[*http.Request]*RequestData
+}
+
+// Inside the WithPerRequestTimeout filetr
+
+func(w http.ResponseWriter, req *http.Request) {
+...
+	data := &RequestData{StartedAt: startedAt, DeadlineAt: deadlineAt, RequestURI: "", AuditID: ""}
+	tracker.Enter(req, data)
+	defer tracker.Leave(req)
+
+	handler.ServeHTTP(w, req)
+...
+```
+
+The `PostStartHook` goroutine will wake up every `5m`, and perform the following operation:
+
+- Iterate through the map entries 
+- If an entry is resident for more than `5m` after the specified `DeadlineAt` then
+  it is reported in the metrics/logs
+- If an entry is resident for more than `10m`, we remove it from the hashmap so the
+  list does not grow infinitely.
+  
+
+The memory cost of the above hashmap is to the `O(number of non long-running requests in flight)`.
+  
+One tradeoff of this global hashmap is the locking contention present due to
+in-flight requests trying to access the synchronized map concurrently. But it
+avoids creating a new goroutine for each request.
+
+##### Post-Tomeout Goroutine
+
+Inside the `WithPerRequestTimeout` filter we can kick off a goroutine to monitor
+post-timeout activity, but this would include all non long-running request(s).
+
+Ideally, we want to spin up goroutine(s) for only those request(s) that time out.
+How do we get notified when the `Context` of a request times out, especially when
+the request handler chain is being executed on a single goroutine?
+
+
+
+### Impact on the Client
+
+It's important to understand how the the client is impacted with this feature enabled; 
+upon timeout of a request, what the client sees depends on the following factors:
+
+- a) Protocol of the request, HTTP/1x or HTTP/2.0.
+- b) Whether the client specifies a timeout in the `timeout` parameter of the request.
+- c) Whether the request handler writes to the `ResponseWriter` object by invoking
+  `Write`, before the deadline exceeds.
+- d) Whether the request handler invokes `Flush` on the `ResponseWriter` object, 
+  preceded  by a `Write`, before the deadline exceeds.
+
+If enabling this feature introduces a change in what the client observes, then it should be 
```

Comment:
the behavior of all supported clients (and probably all clients in the last two years) must be sensible.

Reply:
do we need to worry about any other client than http/1x, http/2.0?